server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -312,6 +312,19 @@ static void parseObjectOrNested(DocumentParserContext context) throws IOExceptio
                 parser = context.parser();
             }
             context = context.createNestedContext((NestedObjectMapper) context.parent());
+        } else if (context.storeSourceModeFromIndexSettings() == Mapper.StoreSourceMode.ALL && context.canAddIgnoredField()) {
```

Comment:
I don't understand why this is needed for the singleton case. I thought that store array source would only apply to arrays, do we also need to store single objects?

Reply:
Hm correct, this is not possible actually.. I'll revert it.

It will be added as a singleton option - there are cases where we may want to store the source as-is, e.g. for geo or full object hierarchies. We also have the equivalent for nested - highjacking `store_array_source`:

https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java#L300

Reply:
Btw this is intended to replace `store_array_source` for objects too.